File: sr_utilities/test/test_hand_finder.py
#!/usr/bin/env python3

# Copyright 2011, 2022 Shadow Robot Company Ltd.
#
# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation version 2 of the License.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import
import unittest
import rospy
import rostest
from sr_utilities.hand_finder import HandFinder
joint_names = ["FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
               "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
               "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"]


class TestHandFinder(unittest.TestCase):
    def test_no_hand_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        hand_finder = HandFinder()

        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 0, "correct parameters without a hand")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 0, "correct parameters without a hand")
        self.assertEqual(len(hand_finder.get_hand_joints()), 0, "correct joints without a hand")

    def test_one_hand_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "right")

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "right", "It should be only right mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "rh_", "It should be only rh_ prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("right", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['right']
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("rh_FFJ3", hand_finder.get_hand_joints()["right"], "FFJ3 joint should be in the joints list")

    def test_one_hand_no_mapping_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "")

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "1", "It should be the serial id as mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "rh_", "It should be only rh_ prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        print(hand_finder.get_hand_joints())
        self.assertIn("1", hand_finder.get_hand_joints(), "Serial should be in the joints result")
        joints = hand_finder.get_hand_joints()['1']  # use serial
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("rh_FFJ3", hand_finder.get_hand_joints()["1"], "FFJ3 joint should be in the joints list")

    def test_one_hand_no_prefix_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "")
        rospy.set_param("hand/mapping/1", "rh")

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "rh", "It should be only right mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "", "It should be only an empty prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("rh", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['rh']
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("FFJ3", hand_finder.get_hand_joints()["rh"], "FFJ3 joint should be in the joints list")

    def test_one_hand_no_prefix_no_mapping_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "")
        rospy.set_param("hand/mapping/1", "")

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "1", "It should be the serial id as mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "", "It should be only an empty prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("1", hand_finder.get_hand_joints(), "Serial should be in the joints result")
        joints = hand_finder.get_hand_joints()['1']
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("FFJ3", hand_finder.get_hand_joints()["1"], "FFJ3 joint should be in the joints list")

    def test_two_hand_no_robot_description_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "right")
        rospy.set_param("hand/joint_prefix/2", "lh_")
        rospy.set_param("hand/mapping/2", "left")

        hand_finder = HandFinder()

        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 2, "It should be two mappings")
        self.assertIn("right", list(hand_finder.get_hand_parameters().mapping.values()), "It should be right mapping")
        self.assertIn("left", list(hand_finder.get_hand_parameters().mapping.values()), "It should be left mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 2, "It should be two joint_prefixes")

        self.assertIn("rh_", list(hand_finder.get_hand_parameters().joint_prefix.values()), "It should be rh_ prefix")
        self.assertIn("lh_", list(hand_finder.get_hand_parameters().joint_prefix.values()), "It should be rh_ prefix")

        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 2, "It should be two mappings")
        self.assertIn("right", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['right']
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("rh_FFJ3", hand_finder.get_hand_joints()["right"], "FFJ3 joint should be in the joints list")
        self.assertIn("left", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['left']
        self.assertEqual(len(joints), 24, "Joint number should be 24")
        self.assertIn("lh_FFJ1", hand_finder.get_hand_joints()["left"], "FFJ1 joint should be in the joints list")

    def test_no_hand_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("robot_description", rospy.get_param("two_hands_description"))

        hand_finder = HandFinder()

        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 0, "correct parameters without a hand")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 0, "correct parameters without a hand")
        self.assertEqual(len(hand_finder.get_hand_joints()), 0, "correct joints without a hand")

    def test_one_hand_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "right")
        rospy.set_param("robot_description", rospy.get_param("right_hand_description"))

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "right", "It should be only right mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "rh_", "It should be only rh_ prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("right", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['right']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("rh_FFJ3", hand_finder.get_hand_joints()["right"],
                         "FFJ3 joint should not be in the joints list")
        self.assertIn("rh_RFJ4", hand_finder.get_hand_joints()["right"], "RFJ4 joint should be in the joints list")

    def test_one_hand_no_mapping_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "")
        rospy.set_param("robot_description", rospy.get_param("right_hand_description"))

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "1", "It should be the serial id as mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "rh_", "It should be only rh_ prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("1", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['1']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("rh_FFJ3", hand_finder.get_hand_joints()["1"], "FFJ3 joint should not be in the joints list")
        self.assertIn("rh_RFJ4", hand_finder.get_hand_joints()["1"], "RFJ4 joint should be in the joints list")

    def test_one_hand_no_prefix_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "")
        rospy.set_param("hand/mapping/1", "rh")
        rospy.set_param("robot_description", rospy.get_param("right_hand_description_no_prefix"))

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "rh", "It should be only right mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "", "It should be only an empty prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("rh", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['rh']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("FFJ3", hand_finder.get_hand_joints()["rh"], "FFJ3 joint should not be in the joints list")
        self.assertIn("RFJ4", hand_finder.get_hand_joints()["rh"], "RFJ4 joint should be in the joints list")

    def test_one_hand_no_prefix_no_ns_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "")
        rospy.set_param("hand/mapping/1", "")
        rospy.set_param("robot_description", rospy.get_param("right_hand_description_no_prefix"))

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 1, "It should be only one mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 1, "It should be only one joint_prefix")
        self.assertEqual(hand_finder.get_hand_parameters().mapping['1'], "1", "It should be the serial id as mapping")
        self.assertEqual(hand_finder.get_hand_parameters().joint_prefix['1'], "", "It should be only an empty prefix")
        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 1, "It should be only one mapping")
        self.assertIn("1", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['1']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("FFJ3", hand_finder.get_hand_joints()["1"], "FFJ3 joint should not be in the joints list")
        self.assertIn("RFJ4", hand_finder.get_hand_joints()["1"], "RFJ4 joint should be in the joints list")

    def test_two_hand_robot_description_exists_finder(self):
        if rospy.has_param("hand"):
            rospy.delete_param("hand")

        if rospy.has_param("robot_description"):
            rospy.delete_param("robot_description")

        rospy.set_param("hand/joint_prefix/1", "rh_")
        rospy.set_param("hand/mapping/1", "right")
        rospy.set_param("hand/joint_prefix/2", "lh_")
        rospy.set_param("hand/mapping/2", "left")
        rospy.set_param("robot_description", rospy.get_param("two_hands_description"))

        hand_finder = HandFinder()
        # hand params
        self.assertIsNotNone(hand_finder.get_hand_parameters(), "Parameters extracted.")
        self.assertEqual(len(hand_finder.get_hand_parameters().mapping), 2, "It should be two mappings")
        self.assertIn("right", list(hand_finder.get_hand_parameters().mapping.values()), "It should be right mapping")
        self.assertIn("left", list(hand_finder.get_hand_parameters().mapping.values()), "It should be left mapping")
        self.assertEqual(len(hand_finder.get_hand_parameters().joint_prefix), 2, "It should be two joint_prefixes")

        self.assertIn("rh_", list(hand_finder.get_hand_parameters().joint_prefix.values()), "It should be rh_ prefix")
        self.assertIn("lh_", list(hand_finder.get_hand_parameters().joint_prefix.values()), "It should be rh_ prefix")

        # hand joints
        self.assertIsNotNone(hand_finder.get_hand_joints(), "Joints extracted.")
        self.assertEqual(len(list(hand_finder.get_hand_joints().keys())), 2, "It should be two mappings")
        self.assertIn("right", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['right']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("rh_FFJ3", hand_finder.get_hand_joints()["right"],
                         "rh_FFJ3 joint should not be in the joints list")
        self.assertIn("rh_RFJ4", hand_finder.get_hand_joints()["right"], "rh_RFJ4 joint should be in the joints list")
        self.assertIn("left", hand_finder.get_hand_joints(), "Mapping should be in the joints result")
        joints = hand_finder.get_hand_joints()['left']
        self.assertEqual(len(joints), 1, "Joint number should be 1")
        self.assertNotIn("lh_FFJ3", hand_finder.get_hand_joints()["left"],
                         "lh_FFJ3 joint should not be in the joints list")
        self.assertIn("lh_LFJ4", hand_finder.get_hand_joints()["left"], "lh_LFJ4 joint should be in the joints list")


if __name__ == "__main__":
    rospy.init_node("test_hand_finder")
    rostest.rosrun("sr_utilities", "test_hand_finder", TestHandFinder)

File: sr_utilities/test/test_hand_finder.cpp
/*
* Copyright 2011 Shadow Robot Company Ltd.
*
* This program is free software: you can redistribute it and/or modify it
* under the terms of the GNU General Public License as published by the Free
* Software Foundation version 2 of the License.
*
* This program is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
* more details.
*
* You should have received a copy of the GNU General Public License along
* with this program. If not, see <http://www.gnu.org/licenses/>.
*/

/*
 * File:  test_hand_finder.cpp
 * Author: Vahid Aminzadeh <vahid@shadowrobot.com>
 * Copyright:
 *
 * @brief see README.md
 */
#include <gtest/gtest.h>
#include "ros/ros.h"
#include <ros/package.h>
#include <utility>
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include "sr_utilities/sr_hand_finder.hpp"
#include <boost/foreach.hpp>

using std::vector;
using std::map;
using std::string;
using std::pair;

void ASSERT_MAP_HAS_VALUE(const map<string, string> &map, const string &value)
{
  bool found = false;
  pair<string, string> item;
  BOOST_FOREACH(item, map)
  {
    if (value == item.second)
    {
      found = true;
    }
  }

  ASSERT_TRUE(found);
}

TEST(SrHandFinder, hand_absent_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  shadow_robot::SrHandFinder hand_finder;
  map<string, vector<string> > hand_joints(hand_finder.get_joints());
  ASSERT_EQ(hand_joints.size(), 0);
}

TEST(SrHandFinder, one_hand_no_robot_description_finder_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  ros::param::set("hand/mapping/1", "right");
  ros::param::set("hand/joint_prefix/1", "rh_");

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  shadow_robot::SrHandFinder hand_finder;
  // hand parameters
  ASSERT_GT(hand_finder.get_hand_parameters().mapping_.size(), 0);
  ASSERT_GT(hand_finder.get_hand_parameters().joint_prefix_.size(), 0);
  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_["1"], "right");
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_["1"], "rh_");

  // hand joints
  ASSERT_EQ(hand_finder.get_joints().size(), 1);
  ASSERT_GT(hand_finder.get_joints().count("right"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("right");
  ASSERT_EQ(rh_joints.size(), 24);
  for (size_t i = 0; i < rh_joints.size(); ++i)
  {
    ROS_DEBUG_STREAM(rh_joints[i]);
    ASSERT_STREQ(rh_joints[i].c_str(), ("rh_" + joint_names[i]).c_str());
  }
}

TEST(SrHandFinder, one_hand_no_mapping_no_robot_description_finder_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  ros::param::set("hand/mapping/1", "");
  ros::param::set("hand/joint_prefix/1", "rh_");

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  shadow_robot::SrHandFinder hand_finder;
  // hand parameters
  ASSERT_GT(hand_finder.get_hand_parameters().mapping_.size(), 0);
  ASSERT_GT(hand_finder.get_hand_parameters().joint_prefix_.size(), 0);
  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_["1"], "1");
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_["1"], "rh_");

  // hand joints
  ASSERT_EQ(hand_finder.get_joints().size(), 1);
  ASSERT_GT(hand_finder.get_joints().count("1"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("1");
  ASSERT_EQ(rh_joints.size(), 24);
  for (size_t i = 0; i < rh_joints.size(); ++i)
  {
    ROS_DEBUG_STREAM(rh_joints[i]);
    ASSERT_STREQ(rh_joints[i].c_str(), ("rh_" + joint_names[i]).c_str());
  }
}

TEST(SrHandFinder, one_hand_no_prefix_no_robot_description_finder_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  ros::param::set("hand/mapping/1", "rh");
  ros::param::set("hand/joint_prefix/1", "");

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  shadow_robot::SrHandFinder hand_finder;
  // hand parameters
  ASSERT_GT(hand_finder.get_hand_parameters().mapping_.size(), 0);
  ASSERT_GT(hand_finder.get_hand_parameters().joint_prefix_.size(), 0);
  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_["1"], "rh");
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_["1"], "");

  // hand joints
  ASSERT_EQ(hand_finder.get_joints().size(), 1);
  ASSERT_GT(hand_finder.get_joints().count("rh"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("rh");
  ASSERT_EQ(rh_joints.size(), 24);
  for (size_t i = 0; i < rh_joints.size(); ++i)
  {
    ROS_DEBUG_STREAM(rh_joints[i]);
    ASSERT_STREQ(rh_joints[i].c_str(), (joint_names[i]).c_str());
  }
}

TEST(SrHandFinder, one_hand_no_mapping_no_prefix_no_robot_description_finder_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  ros::param::set("hand/mapping/1", "");
  ros::param::set("hand/joint_prefix/1", "");

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  shadow_robot::SrHandFinder hand_finder;
  // hand parameters
  ASSERT_GT(hand_finder.get_hand_parameters().mapping_.size(), 0);
  ASSERT_GT(hand_finder.get_hand_parameters().joint_prefix_.size(), 0);
  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_["1"], "1");
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_["1"], "");

  // hand joints
  ASSERT_EQ(hand_finder.get_joints().size(), 1);
  ASSERT_GT(hand_finder.get_joints().count("1"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("1");
  ASSERT_EQ(rh_joints.size(), 24);
  for (size_t i = 0; i < rh_joints.size(); ++i)
  {
    ROS_DEBUG_STREAM(rh_joints[i]);
    ASSERT_STREQ(rh_joints[i].c_str(), (joint_names[i]).c_str());
  }
}

TEST(SrHandFinder, one_hand_robot_description_exists_finger_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  string right_hand_description;
  ros::param::get("right_hand_description", right_hand_description);
  ros::param::set("robot_description", right_hand_description);

  ros::param::set("hand/mapping/1", "right");
  ros::param::set("hand/joint_prefix/1", "rh_");

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  shadow_robot::SrHandFinder hand_finder;
  // hand parameters
  ASSERT_GT(hand_finder.get_hand_parameters().mapping_.size(), 0);
  ASSERT_GT(hand_finder.get_hand_parameters().joint_prefix_.size(), 0);
  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_["1"], "right");
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_["1"], "rh_");

  // hand joints
  ASSERT_EQ(hand_finder.get_joints().size(), 1);
  ASSERT_GT(hand_finder.get_joints().count("right"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("right");
  ASSERT_EQ(rh_joints.size(), 1);  // only RFJ4 is there
  ASSERT_EQ(std::find(rh_joints.begin(), rh_joints.end(), "rh_FFJ3"), rh_joints.end());
  ASSERT_NE(std::find(rh_joints.begin(), rh_joints.end(), "rh_RFJ4"), rh_joints.end());
}

TEST(SrHandFinder, two_hand_robot_description_exists_finder_test)
{
  if (ros::param::has("hand"))
  {
    ros::param::del("hand");
  }

  if (ros::param::has("robot_description"))
  {
    ros::param::del("robot_description");
  }

  ros::param::set("hand/mapping/1", "right");
  ros::param::set("hand/joint_prefix/1", "rh_");
  ros::param::set("hand/mapping/2", "left");
  ros::param::set("hand/joint_prefix/2", "lh_");

  string two_hands_description;
  ros::param::get("two_hands_description", two_hands_description);
  ros::param::set("robot_description", two_hands_description);

  const string joint_names[] =
  {
    "FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
    "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
    "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"
  };

  unsigned idx = 0;
  const string dir[] =  {"left", "right"};

  shadow_robot::SrHandFinder hand_finder;

  ASSERT_EQ(hand_finder.get_hand_parameters().mapping_.size(), 2);
  ASSERT_EQ(hand_finder.get_hand_parameters().joint_prefix_.size(), 2);
  ASSERT_EQ(hand_finder.get_joints().size(), 2);

  ASSERT_MAP_HAS_VALUE(hand_finder.get_hand_parameters().mapping_, "right");
  ASSERT_MAP_HAS_VALUE(hand_finder.get_hand_parameters().mapping_, "left");

  ASSERT_MAP_HAS_VALUE(hand_finder.get_hand_parameters().joint_prefix_, "rh_");
  ASSERT_MAP_HAS_VALUE(hand_finder.get_hand_parameters().joint_prefix_, "lh_");

  ASSERT_GT(hand_finder.get_joints().count("right"), 0);
  const vector<string> rh_joints = hand_finder.get_joints().at("right");
  ASSERT_EQ(std::find(rh_joints.begin(), rh_joints.end(), "rh_FFJ3"), rh_joints.end());
  ASSERT_NE(std::find(rh_joints.begin(), rh_joints.end(), "rh_RFJ4"), rh_joints.end());

  ASSERT_GT(hand_finder.get_joints().count("left"), 0);
  const vector<string> lh_joints = hand_finder.get_joints().at("left");
  ASSERT_EQ(std::find(lh_joints.begin(), lh_joints.end(), "lh_FFJ1"), lh_joints.end());
  ASSERT_NE(std::find(lh_joints.begin(), lh_joints.end(), "lh_LFJ4"), lh_joints.end());
}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "test_hand_finder");
  ros::NodeHandle nh;
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

